src/components/Trade/TradeInputV2.tsx

```diff
@@ -28,54 +39,91 @@ import { type TradeState, TradeAmountInputField, TradeRoutePaths } from './types
   const quote = useWatch({ control, name: 'quote' })
   const feeAssetFiatRate = useWatch({ control, name: 'feeAssetFiatRate' })
+  const buyAssetFiatRate = useWatch({ control, name: 'buyAssetFiatRate' })
   const fees = useWatch({ control, name: 'fees' })
   const sellAssetAccountId = useWatch({ control, name: 'sellAssetAccountId' })
   const buyAssetAccountId = useWatch({ control, name: 'buyAssetAccountId' })
   const fiatSellAmount = useWatch({ control, name: 'fiatSellAmount' })
   const fiatBuyAmount = useWatch({ control, name: 'fiatBuyAmount' })
   const slippage = useWatch({ control, name: 'slippage' })
 
-  const translate = useTranslate()
-
+  // Selectors
   const sellFeeAsset = useAppSelector(state =>
     selectFeeAssetById(state, sellTradeAsset?.asset?.assetId ?? ethAssetId),
   )
-  const sellAssetBalance = useAppSelector(state =>
+  const feeAssetBalance = useAppSelector(state =>
+    sellFeeAsset
+      ? selectPortfolioCryptoHumanBalanceByAssetId(state, { assetId: sellFeeAsset?.assetId })
+      : null,
+  )
+  const sellAssetBalanceCrypto = useAppSelector(state =>
     selectPortfolioCryptoBalanceByFilter(state, {
       accountId: sellAssetAccountId,
       assetId: sellTradeAsset?.asset?.assetId ?? '',
     }),
   )
 
-  const protocolFee = fromBaseUnit(bnOrZero(fees?.tradeFee), buyTradeAsset?.asset?.precision ?? 0)
-  const toCryptoAmountAfterFees = bnOrZero(buyTradeAsset?.amount).minus(bnOrZero(protocolFee))
+  const sellAssetBalanceHuman = useAppSelector(state =>
+    selectPortfolioCryptoHumanBalanceByFilter(state, {
+      accountId: sellAssetAccountId,
+      assetId: sellTradeAsset?.asset?.assetId ?? '',
+    }),
+  )
+
+  // Constants
+  const sellAssetAccountSupported = (() => {
+    const sellAsset = sellTradeAsset?.asset
+    if (sellAsset && isSupportedNonUtxoSwappingChain(sellAsset.chainId)) return true
+    if (sellAsset && isSupportedUtxoSwappingChain(sellAsset.chainId) && sellAssetAccountId) {
+      const { utxoParams } = getUtxoParams(sellAssetAccountId)
+      return !!utxoParams?.bip44Params
+    }
+    return false
+  })()
+
+  const protocolFeeCrypto = bnOrZero(fees?.tradeFee).div(bnOrZero(buyAssetFiatRate)).toString()
+  const toCryptoAmountBeforeFees = bnOrZero(buyTradeAsset?.amount).plus(bnOrZero(protocolFeeCrypto))
   const gasFee = bnOrZero(fees?.fee).times(bnOrZero(feeAssetFiatRate)).toString()
+  const hasValidSellAmount = bnOrZero(sellTradeAsset?.amount).gt(0)
 
-  const handleInputChange = (action: TradeAmountInputField, amount: string) => {
+  const handleInputChange = async (action: TradeAmountInputField, amount: string) => {
     setValue('amount', amount)
     setValue('action', action)
-    setTradeAmounts({ amount, action })
+
+    isLoadingFiatRateData || isLoadingTradeQuote
+      ? await setTradeAmountsSynchronous({ amount, action })
+      : setTradeAmountsAsynchronous({ amount, action })
   }
 
-  const handleToggle = () => {
+  const handleToggle = async () => {
```

Comment:
Can we `useCallback` this while at it? 🙏🏽

Reply:
Addressed in https://github.com/shapeshift/web/pull/2789/commits/abceb92a5d2fe0f2b312af7aa7a1b0759ed136ff